darwin/DarwinProcess.c

```diff
@@ -327,6 +320,25 @@ void DarwinProcess_setFromKInfoProc(Process* proc, const struct kinfo_proc* ps,
       }
    }
 
+   if (proc->tty_name == NULL && (dev_t)proc->tty_nr != NODEV) {
+      /* The call to devname() is extremely expensive (due to lstat)
+       * and represents ~95% of htop's CPU usage when there is high
+       * process turnover.
+       *
+       * To mitigate this we only fetch TTY information if the TTY
+       * field is enabled in the settings.
+       */
+      if (proc->settings->ss->flags & PROCESS_FLAG_TTY) {
+         const char *name = devname(proc->tty_nr, S_IFCHR);
+         if (name) {
+            proc->tty_name = xStrndup(name, 512);
+         } else {
+            /* devname failed: prevent us from calling it again */
+            proc->tty_nr = NODEV;
+         }
+      }
+   }
+
```

Comment:
By moving this code block from where it was before you are checking for a terminal on each process information update compared to just when the process was freshly created. And as `proc->tty_nr` doesn't change after the first initialization above this move does not make much sense.

Reply:
I moved it out of the `!exists` block so that changes to the settings (enabling/disabling the TTY column) will be reflected on each update. Otherwise, if the TTY column is enabled, after `htop` has been running, it will only be shown for processes that are created post-enablement (previously existing processes will not have the TTY field populated).

Also, the purpose of this block is lazily setting the `proc->tty_name` field. We only modify the `proc->tty_nr` if there is an error calling `devname()` and do so to prevent us from calling it again (on the assumption that if it failed once for a given `dev` number it'll continue to fail given that `dev`).